File: ios/HybridFrameGrab.swift
//
//  HybridFrameGrab.swift
//  NitroFrameGrab
//
//  video URI -> AVAssetImageGenerator -> bounded scale -> ImageIO JPEG -> rename(2)
//

// AVFoundation is a pre-concurrency framework: the iOS 15 fallback path below
// hands `AVAsynchronousKeyValueLoading` to a `@Sendable` completion handler,
// which it has no Sendable annotations for.
@preconcurrency import AVFoundation
import CoreGraphics
import CoreMedia
import Foundation
import ImageIO
import NitroModules
import UniformTypeIdentifiers

private let maxOutputArea = 16_777_216
private let minMaxWidth = 1
private let maxMaxWidth = 4096

/// Blocking filesystem/ImageIO work runs here so it never occupies a slot in
/// Swift's cooperative thread pool (which is sized for non-blocking work).
/// Only jobs that already hold a scheduler slot reach this queue.
private let frameGrabIOQueue = DispatchQueue(
  label: "com.lordronz.framegrab.io", qos: .userInitiated, attributes: .concurrent)

/// Pre-admission validation stats the filesystem, which must not happen on the
/// JS thread either. Serial on purpose: a burst of rejected requests should not
/// explode GCD's thread pool before any of them has a slot.
private let frameGrabIntakeQueue = DispatchQueue(
  label: "com.lordronz.framegrab.intake", qos: .userInitiated)

private func onQueue<T>(
  _ queue: DispatchQueue, _ body: @escaping @Sendable () throws -> T
) async throws -> T {
  try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<T, Error>) in
    queue.async {
      do {
        continuation.resume(returning: try body())
      } catch {
        continuation.resume(throwing: error)
      }
    }
  }
}

private func millis(since start: DispatchTime) -> Double {
  Double(DispatchTime.now().uptimeNanoseconds - start.uptimeNanoseconds) / 1_000_000
}

final class HybridFrameGrab: HybridFrameGrabSpec {
  var maxConcurrency: Double {
    get { Double(FrameGrabScheduler.shared.maxConcurrency) }
    set { FrameGrabScheduler.shared.maxConcurrency = Int(newValue.rounded()) }
  }

  func extract(request: NativeFrameGrabRequest) throws -> Promise<NativeFrameGrabOutcome> {
    // Copy out of the C++-backed struct so the task captures only Swift values.
    let sourceUri = request.sourceUri
    let destinationUri = request.destinationUri
    let timeUs = request.timeUs
    let maxWidth = request.maxWidth
    let quality = request.quality
    let precise = request.precise

    return Promise.async {
      do {
        let success = try await FrameGrabJob.run(
          sourceUri: sourceUri,
          destinationUri: destinationUri,
          timeUs: timeUs,
          maxWidth: maxWidth,
          quality: quality,
          precise: precise)
        return NativeFrameGrabOutcome(result: success, error: nil)
      } catch let failure as FrameGrabFailure {
        return NativeFrameGrabOutcome(
          result: nil,
          error: NativeFrameGrabFailure(code: failure.code.rawValue, message: failure.message))
      } catch {
        return NativeFrameGrabOutcome(
          result: nil,
          error: NativeFrameGrabFailure(
            code: FrameGrabCode.internalError.rawValue,
            message: "Unexpected native failure: \(error.localizedDescription)"))
      }
    }
  }
}

// MARK: - Job

private enum FrameGrabJob {

  static func run(
    sourceUri: String,
    destinationUri: String,
    timeUs: Double,
    maxWidth: Double,
    quality: Double,
    precise: Bool
  ) async throws -> NativeFrameGrabSuccess {

    // 1. Re-validate scalars at the native boundary before any conversion.
    guard timeUs.isFinite, timeUs >= 0, timeUs <= 9_007_199_254_740_991,
      timeUs.rounded() == timeUs
    else {
      throw FrameGrabFailure(.invalidArgument, "`timeMs` is out of the representable range.")
    }
    guard maxWidth.isFinite, maxWidth.rounded() == maxWidth,
      Int(maxWidth) >= minMaxWidth, Int(maxWidth) <= maxMaxWidth
    else {
      throw FrameGrabFailure(
        .invalidArgument, "`maxWidth` must be an integer in \(minMaxWidth)...\(maxMaxWidth).")
    }
    guard quality.isFinite, quality >= 0, quality <= 1 else {
      throw FrameGrabFailure(.invalidArgument, "`quality` must be between 0 and 1.")
    }
    let timeUsInt = Int64(timeUs)
    let widthBound = Int(maxWidth)

    // 2. Cheap URI + destination checks, off the cooperative pool.
    let source = try FrameGrabPaths.resolveSource(sourceUri)
    let destination = try FrameGrabPaths.resolveDestination(destinationUri)
    try await onQueue(frameGrabIntakeQueue) {
      try FrameGrabPaths.validateDestination(destination)
      try FrameGrabPaths.assertDistinct(source: source, destination: destination)
      if case .localFile(let url) = source,
        !FileManager.default.fileExists(atPath: url.path)
      {
        throw FrameGrabFailure(
          .sourceUnreadable, "Source file does not exist: \(frameGrabRedact(sourceUri))")
      }
    }

    // 3. Reserve the destination and take a slot. Nothing native is opened before this.
    let key = FrameGrabPaths.reservationKey(for: destination)
    let queueWaitMs = try await FrameGrabScheduler.shared.acquire(key: key)
    defer { FrameGrabScheduler.shared.release(key: key) }

    let temporaryURL = destination.deletingLastPathComponent()
      .appendingPathComponent(".framegrab-\(UUID().uuidString).jpg.tmp")
    var committed = false
    defer {
      if !committed {
        try? FileManager.default.removeItem(at: temporaryURL)
      }
    }

    // 4. Source + metadata.
    let sourceStart = DispatchTime.now()
    let asset = AVURLAsset(url: source.url)
    let sourceMs = millis(since: sourceStart)

    let metadataStart = DispatchTime.now()
    let (track, duration) = try await loadVideoTrack(asset, sourceUri: sourceUri)
    let display = try await displaySize(of: track)

    guard display.width.isFinite, display.height.isFinite,
      display.width >= 1, display.height >= 1
    else {
      throw FrameGrabFailure(
        .sourceUnreadable, "Video track reports unusable dimensions.")
    }

    // A timestamp equal to the duration is out of range. An unknown or
    // indefinite duration is not proof of anything: let extraction decide.
    let durationSeconds = CMTimeGetSeconds(duration)
    if duration.isValid, !duration.isIndefinite, durationSeconds.isFinite, durationSeconds > 0 {
      let durationUs = durationSeconds * 1_000_000
      if Double(timeUsInt) >= durationUs {
        throw FrameGrabFailure(
          .timestampOutOfRange,
          "Requested \(Double(timeUsInt) / 1000) ms but the source is \(durationUs / 1000) ms long.")
      }
    }

    let (targetWidth, targetHeight) = try fit(
      width: display.width, height: display.height, maxWidth: widthBound)
    let metadataMs = millis(since: metadataStart)

    // 5. Extraction.
    let extractStart = DispatchTime.now()
    let generator = AVAssetImageGenerator(asset: asset)
    generator.appliesPreferredTrackTransform = true
    generator.apertureMode = .cleanAperture
    generator.maximumSize = CGSize(width: targetWidth, height: targetHeight)
    let tolerance: CMTime = precise ? .zero : .positiveInfinity
    generator.requestedTimeToleranceBefore = tolerance
    generator.requestedTimeToleranceAfter = tolerance

    let requestedTime = CMTime(value: timeUsInt, timescale: 1_000_000)
    let image = try await generateImage(generator: generator, at: requestedTime)
    let extractMs = millis(since: extractStart)

    guard image.width > 0, image.height > 0 else {
      throw FrameGrabFailure(.frameExtraction, "Decoder produced an empty image.")
    }

    // 6. Encode straight to the temporary sibling; the encoder streams to disk,
    //    so encode and write are one measurement.
    let encodeStart = DispatchTime.now()
    let boxedImage = FrameGrabBox(image)
    let boxedTemporary = FrameGrabBox(temporaryURL)
    let encoded = try await onQueue(frameGrabIOQueue) { () -> (Int, Int, UInt64) in
      let final = try bound(boxedImage.value, maxWidth: widthBound)
      guard final.width * final.height <= maxOutputArea else {
        throw FrameGrabFailure(
          .internalError,
          "Generated image (\(final.width)x\(final.height)) exceeds the output area limit.")
      }
      try encodeJPEG(final, to: boxedTemporary.value, quality: quality)
      let attributes = try FileManager.default.attributesOfItem(
        atPath: boxedTemporary.value.path)
      guard let size = (attributes[.size] as? NSNumber)?.uint64Value, size > 0 else {
        throw FrameGrabFailure(.encodeFailed, "Encoder produced an empty file.")
      }
      return (final.width, final.height, size)
    }
    let (width, height, byteSize) = encoded
    let encodeMs = millis(since: encodeStart)

    // 7. Commit. After this point the old destination is gone and we never
    //    fail the request for bookkeeping reasons.
    let finalizeStart = DispatchTime.now()
    let boxedDestination = FrameGrabBox(destination)
    try await onQueue(frameGrabIOQueue) {
      try atomicReplace(from: boxedTemporary.value, to: boxedDestination.value)
    }
    committed = true
    let finalizeMs = millis(since: finalizeStart)

    return NativeFrameGrabSuccess(
      uri: destination.absoluteString,
      width: Double(width),
      height: Double(height),
      size: Double(byteSize),
      timings: NativeFrameGrabTimings(
        queueWaitMs: queueWaitMs,
        sourceMs: sourceMs,
        metadataMs: metadataMs,
        extractMs: extractMs,
        encodeMs: encodeMs,
        finalizeMs: finalizeMs))
  }

  // MARK: Metadata

  private static func loadVideoTrack(
    _ asset: AVURLAsset, sourceUri: String
  ) async throws -> (AVAssetTrack, CMTime) {
    if #available(iOS 16.0, tvOS 16.0, visionOS 1.0, *) {
      do {
        async let durationTask = asset.load(.duration)
        async let tracksTask = asset.loadTracks(withMediaType: .video)
        let (duration, tracks) = try await (durationTask, tracksTask)
        guard let track = tracks.first else {
          throw FrameGrabFailure(
            .sourceUnreadable, "No usable video track in \(frameGrabRedact(sourceUri)).")
        }
        return (track, duration)
      } catch let failure as FrameGrabFailure {
        throw failure
      } catch {
        throw FrameGrabFailure(
          .sourceUnreadable,
          "Could not read \(frameGrabRedact(sourceUri)): \(error.localizedDescription)")
      }
    }
    try await loadLegacyValues(asset, keys: ["duration", "tracks"], sourceUri: sourceUri)
    guard let track = asset.tracks(withMediaType: .video).first else {
      throw FrameGrabFailure(
        .sourceUnreadable, "No usable video track in \(frameGrabRedact(sourceUri)).")
    }
    return (track, asset.duration)
  }

  private static func displaySize(of track: AVAssetTrack) async throws -> CGSize {
    let naturalSize: CGSize
    let transform: CGAffineTransform
    let formats: [CMFormatDescription]

    if #available(iOS 16.0, tvOS 16.0, visionOS 1.0, *) {
      do {
        let loaded = try await track.load(.naturalSize, .preferredTransform, .formatDescriptions)
        naturalSize = loaded.0
        transform = loaded.1
        formats = loaded.2
      } catch {
        throw FrameGrabFailure(
          .sourceUnreadable, "Could not read track properties: \(error.localizedDescription)")
      }
    } else {
      try await loadLegacyValues(
        track, keys: ["naturalSize", "preferredTransform", "formatDescriptions"],
        sourceUri: "")
      naturalSize = track.naturalSize
      transform = track.preferredTransform
      // The legacy accessor is an untyped `NSArray`. Cast the array, not each
      // element: a per-element `as?` to a CoreFoundation type is a hard error
      // ("will always succeed"), because CF types bridge unconditionally.
      formats = track.formatDescriptions as? [CMFormatDescription] ?? []
    }

    // Prefer presentation dimensions: they fold in the pixel aspect ratio and
    // clean aperture, which `naturalSize` alone ignores.
    var base = naturalSize
    if let format = formats.first {
      // Returns CMVideoDimensions (Int32), not a CGSize.
      let presentation = CMVideoFormatDescriptionGetPresentationDimensions(
        format, usePixelAspectRatio: true, useCleanAperture: true)
      if presentation.width > 0, presentation.height > 0 {
        base = CGSize(
          width: CGFloat(presentation.width), height: CGFloat(presentation.height))
      }
    }

    // `preferredTransform` carries the display rotation; take the bounding box.
    let rotated = CGRect(origin: .zero, size: base).applying(transform)
    return CGSize(width: abs(rotated.width), height: abs(rotated.height))
  }

  @available(iOS, deprecated: 16.0)
  private static func loadLegacyValues(
    _ object: AVAsynchronousKeyValueLoading, keys: [String], sourceUri: String
  ) async throws {
    try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<Void, Error>) in
      object.loadValuesAsynchronously(forKeys: keys) {
        for key in keys {
          var error: NSError?
          let status = object.statusOfValue(forKey: key, error: &error)
          guard status == .loaded else {
            continuation.resume(
              throwing: FrameGrabFailure(
                .sourceUnreadable,
                "Could not load \"\(key)\" \(sourceUri.isEmpty ? "" : "for \(frameGrabRedact(sourceUri)) ")"
                  + "(\(error?.localizedDescription ?? "status \(status.rawValue)"))."))
            return
          }
        }
        continuation.resume()
      }
    }
  }

  // MARK: Scaling

  /// Fits the display size inside `maxWidth` without ever upscaling.
  private static func fit(width: CGFloat, height: CGFloat, maxWidth: Int) throws -> (Int, Int) {
    let scale = min(1.0, Double(maxWidth) / Double(width))
    let scaledWidth = Int((Double(width) * scale).rounded())
    let scaledHeight = Int((Double(height) * scale).rounded())
    let targetWidth = max(1, min(scaledWidth, maxWidth))
    let targetHeight = max(1, scaledHeight)
    guard targetWidth * targetHeight <= maxOutputArea else {
      throw FrameGrabFailure(
        .invalidArgument,
        "Output would be \(targetWidth)x\(targetHeight), over the \(maxOutputArea) pixel limit.")
    }
    return (targetWidth, targetHeight)
  }

  // MARK: Extraction

  private static func generateImage(
    generator: AVAssetImageGenerator, at time: CMTime
  ) async throws -> CGImage {
    if #available(iOS 16.0, tvOS 16.0, visionOS 1.0, *) {
      do {
        return try await generator.image(at: time).image
      } catch let failure as FrameGrabFailure {
        throw failure
      } catch {
        throw FrameGrabFailure(
          .frameExtraction, "Frame extraction failed: \(error.localizedDescription)")
      }
    }
    return try await withCheckedThrowingContinuation {
      (continuation: CheckedContinuation<CGImage, Error>) in
      // With a single requested time this handler runs exactly once.
      generator.generateCGImagesAsynchronously(forTimes: [NSValue(time: time)]) {
        _, image, _, result, error in
        switch result {
        case .succeeded:
          if let image {
            continuation.resume(returning: image)
          } else {
            continuation.resume(
              throwing: FrameGrabFailure(.frameExtraction, "Decoder returned no image."))
          }
        case .failed:
          continuation.resume(
            throwing: FrameGrabFailure(
              .frameExtraction,
              "Frame extraction failed: \(error?.localizedDescription ?? "unknown error")"))
        case .cancelled:
          continuation.resume(
            throwing: FrameGrabFailure(.frameExtraction, "Frame extraction was cancelled."))
        @unknown default:
          continuation.resume(
            throwing: FrameGrabFailure(.internalError, "Unknown image generation result."))
        }
      }
    }
  }

  /// Single bounded, aspect-preserving redraw, based on the image we actually got.
  ///
  /// `maximumSize` should already have handled this, and normally this is a
  /// no-op. The contract says the final width never exceeds `maxWidth`, so if the
  /// generator ever rounds the other way we honour the contract instead of
  /// failing an otherwise valid request.
  private static func bound(_ image: CGImage, maxWidth: Int) throws -> CGImage {
    guard image.width > maxWidth else { return image }

    let scale = Double(maxWidth) / Double(image.width)
    let height = max(1, Int((Double(image.height) * scale).rounded()))
    guard
      let context = CGContext(
        data: nil,
        width: maxWidth,
        height: height,
        bitsPerComponent: 8,
        bytesPerRow: 0,
        space: CGColorSpaceCreateDeviceRGB(),
        // JPEG has no alpha channel; dropping it here avoids a needless convert.
        bitmapInfo: CGImageAlphaInfo.noneSkipLast.rawValue)
    else {
      throw FrameGrabFailure(.encodeFailed, "Could not allocate a scaling context.")
    }
    context.interpolationQuality = .high
    context.draw(image, in: CGRect(x: 0, y: 0, width: maxWidth, height: height))
    guard let scaled = context.makeImage() else {
      throw FrameGrabFailure(.encodeFailed, "Scaling the extracted frame failed.")
    }
    return scaled
  }

  // MARK: Encode + commit

  private static func encodeJPEG(_ image: CGImage, to url: URL, quality: Double) throws {
    guard
      let destination = CGImageDestinationCreateWithURL(
        url as CFURL, UTType.jpeg.identifier as CFString, 1, nil)
    else {
      throw FrameGrabFailure(
        .destinationWrite, "Could not open a temporary file for writing in the destination folder.")
    }
    // Pixels are already upright (appliesPreferredTrackTransform), so no EXIF
    // orientation tag is written: consumers get what they see.
    CGImageDestinationAddImage(
      destination, image,
      [kCGImageDestinationLossyCompressionQuality: quality] as CFDictionary)
    guard CGImageDestinationFinalize(destination) else {
      throw FrameGrabFailure(.encodeFailed, "JPEG encoding failed.")
    }
  }

  /// `rename(2)` on the same filesystem is atomic and replaces an existing
  /// destination in one step. The temporary file is a sibling, so EXDEV
  /// cannot happen. Never delete-then-move: that loses the old file on failure.
  private static func atomicReplace(from temporary: URL, to destination: URL) throws {
    var savedErrno: Int32 = 0
    let ok = temporary.path.withCString { source in
      destination.path.withCString { target -> Bool in
        if rename(source, target) == 0 { return true }
        savedErrno = errno
        return false
      }
    }
    guard ok else {
      throw FrameGrabFailure(
        .destinationWrite,
        "Could not finalize the destination file: \(String(cString: strerror(savedErrno)))")
    }
  }
}

File: android/src/main/java/com/margelo/nitro/framegrab/HybridFrameGrab.kt
package com.margelo.nitro.framegrab

import android.graphics.Bitmap
import android.graphics.Matrix
import android.media.MediaMetadataRetriever
import android.net.Uri
import android.os.Build
import androidx.annotation.Keep
import com.facebook.proguard.annotations.DoNotStrip
import com.margelo.nitro.NitroModules
import com.margelo.nitro.core.Promise
import java.io.BufferedOutputStream
import java.io.File
import java.io.FileOutputStream
import kotlin.math.max
import kotlin.math.min
import kotlin.math.roundToInt

private const val MAX_OUTPUT_AREA = 16_777_216
private const val MIN_MAX_WIDTH = 1
private const val MAX_MAX_WIDTH = 4096

/**
 * video URI -> MediaMetadataRetriever -> bounded scale -> Bitmap.compress -> rename(2)
 */
// JNI constructs this class by name; keep annotations on the spec are not inherited.
@Keep
@DoNotStrip
class HybridFrameGrab : HybridFrameGrabSpec() {
    override var maxConcurrency: Double
        get() = FrameGrabDispatcher.maxConcurrency.toDouble()
        set(value) {
            FrameGrabDispatcher.maxConcurrency = value.roundToInt()
        }

    override fun extract(request: NativeFrameGrabRequest): Promise<NativeFrameGrabOutcome> {
        val promise = Promise<NativeFrameGrabOutcome>()

        // Returns to JS immediately; nothing below runs on the JS thread.
        FrameGrabDispatcher.onIntake {
            try {
                val plan = prepare(request)
                FrameGrabDispatcher.submit(plan.reservationKey) { queueWaitMs ->
                    promise.resolve(runJob(plan, queueWaitMs))
                }
            } catch (e: FrameGrabException) {
                promise.resolve(failure(e.code, e.message ?: "Extraction failed."))
            } catch (e: Throwable) {
                promise.resolve(
                    failure(FrameGrabCode.INTERNAL, "Unexpected native failure: $e"),
                )
            }
        }

        return promise
    }

    // region Intake

    private class Plan(
        val source: FrameGrabSource,
        val destination: File,
        val reservationKey: String,
        val timeUs: Long,
        val maxWidth: Int,
        val quality: Int,
        val precise: Boolean,
    )

    /**
     * Cheap validation and path resolution. Runs before the reservation is taken
     * and before any native media object exists.
     */
    private fun prepare(request: NativeFrameGrabRequest): Plan {
        val timeUs = request.timeUs
        if (!timeUs.isFinite() || timeUs < 0 || timeUs > 9_007_199_254_740_991.0 ||
            timeUs != Math.floor(timeUs)
        ) {
            throw FrameGrabException(
                FrameGrabCode.INVALID_ARGUMENT,
                "`timeMs` is out of the representable range.",
            )
        }
        val maxWidth = request.maxWidth
        if (!maxWidth.isFinite() || maxWidth != Math.floor(maxWidth) ||
            maxWidth < MIN_MAX_WIDTH || maxWidth > MAX_MAX_WIDTH
        ) {
            throw FrameGrabException(
                FrameGrabCode.INVALID_ARGUMENT,
                "`maxWidth` must be an integer in $MIN_MAX_WIDTH..$MAX_MAX_WIDTH.",
            )
        }
        val quality = request.quality
        if (!quality.isFinite() || quality < 0.0 || quality > 1.0) {
            throw FrameGrabException(
                FrameGrabCode.INVALID_ARGUMENT,
                "`quality` must be between 0 and 1.",
            )
        }

        val source = FrameGrabPaths.resolveSource(request.sourceUri)
        val destination = FrameGrabPaths.resolveDestination(request.destinationUri)
        FrameGrabPaths.validateDestination(destination)
        FrameGrabPaths.assertDistinct(source, destination, NitroModules.applicationContext)

        if (source is FrameGrabSource.LocalFile && !source.file.isFile) {
            throw FrameGrabException(
                FrameGrabCode.SOURCE_UNREADABLE,
                "Source file does not exist: ${redact(request.sourceUri)}",
            )
        }

        return Plan(
            source = source,
            destination = destination,
            reservationKey = FrameGrabPaths.reservationKey(destination),
            timeUs = timeUs.toLong(),
            maxWidth = maxWidth.toInt(),
            quality = (quality * 100).roundToInt().coerceIn(0, 100),
            precise = request.precise,
        )
    }

    // endregion

    // region Job

    private fun runJob(
        plan: Plan,
        queueWaitMs: Double,
    ): NativeFrameGrabOutcome {
        val temporary =
            File(
                plan.destination.absoluteFile.parentFile,
                ".framegrab-${java.util.UUID.randomUUID()}.jpg.tmp",
            )
        var committed = false
        var retriever: MediaMetadataRetriever? = null

        try {
            // 1. Source.
            val sourceStart = System.nanoTime()
            retriever = MediaMetadataRetriever()
            openSource(retriever, plan.source)
            val sourceMs = elapsedMs(sourceStart)

            // 2. Metadata + bounds.
            val metadataStart = System.nanoTime()
            val metadata = readMetadata(retriever, plan)
            val (targetWidth, targetHeight) =
                fit(metadata.orientedWidth, metadata.orientedHeight, plan.maxWidth)
            val metadataMs = elapsedMs(metadataStart)

            // 3. Extraction.
            val extractStart = System.nanoTime()
            val option =
                if (plan.precise) {
                    MediaMetadataRetriever.OPTION_CLOSEST
                } else {
                    MediaMetadataRetriever.OPTION_CLOSEST_SYNC
                }
            var bitmap = extractFrame(retriever, plan.timeUs, option, targetWidth, targetHeight)
            val extractMs = elapsedMs(extractStart)

            try {
                // 4. Orientation, then a single bounded resize if the platform
                //    handed us something larger than we asked for.
                val rotated =
                    rotateIfPlatformDidNot(
                        bitmap,
                        metadata.rotationDegrees,
                        metadata.orientedWidth,
                        metadata.orientedHeight,
                    )
                if (rotated !== bitmap) {
                    bitmap.recycle()
                    bitmap = rotated
                }

                val bounded = boundWidth(bitmap, plan.maxWidth)
                if (bounded !== bitmap) {
                    bitmap.recycle()
                    bitmap = bounded
                }

                val width = bitmap.width
                val height = bitmap.height
                if (width <= 0 || height <= 0) {
                    throw FrameGrabException(
                        FrameGrabCode.FRAME_EXTRACTION,
                        "Decoder produced an empty image.",
                    )
                }
                if (width > plan.maxWidth || width.toLong() * height > MAX_OUTPUT_AREA) {
                    throw FrameGrabException(
                        FrameGrabCode.INTERNAL,
                        "Generated image (${width}x$height) exceeds the requested bounds.",
                    )
                }

                // 5. Encode straight into the temporary sibling. `compress`
                //    streams to the stream, so encode and write are one phase.
                val encodeStart = System.nanoTime()
                BufferedOutputStream(FileOutputStream(temporary)).use { out ->
                    val ok = bitmap.compress(Bitmap.CompressFormat.JPEG, plan.quality, out)
                    if (!ok) {
                        throw FrameGrabException(
                            FrameGrabCode.ENCODE_FAILED,
                            "JPEG encoding failed.",
                        )
                    }
                    out.flush()
                }
                val byteSize = temporary.length()
                if (byteSize <= 0L) {
                    throw FrameGrabException(
                        FrameGrabCode.ENCODE_FAILED,
                        "Encoder produced an empty file.",
                    )
                }
                val encodeMs = elapsedMs(encodeStart)

                // 6. Commit. `File.renameTo` is rename(2): on the same filesystem
                //    it atomically replaces the destination. Never delete-then-move.
                val finalizeStart = System.nanoTime()
                if (!temporary.renameTo(plan.destination)) {
                    throw FrameGrabException(
                        FrameGrabCode.DESTINATION_WRITE,
                        "Could not finalize the destination file.",
                    )
                }
                committed = true
                val finalizeMs = elapsedMs(finalizeStart)

                return NativeFrameGrabOutcome(
                    result =
                        NativeFrameGrabSuccess(
                            uri = Uri.fromFile(plan.destination).toString(),
                            width = width.toDouble(),
                            height = height.toDouble(),
                            size = byteSize.toDouble(),
                            timings =
                                NativeFrameGrabTimings(
                                    queueWaitMs = queueWaitMs,
                                    sourceMs = sourceMs,
                                    metadataMs = metadataMs,
                                    extractMs = extractMs,
                                    encodeMs = encodeMs,
                                    finalizeMs = finalizeMs,
                                ),
                        ),
                    error = null,
                )
            } finally {
                bitmap.recycle()
            }
        } catch (e: FrameGrabException) {
            return failure(e.code, e.message ?: "Extraction failed.")
        } catch (e: Throwable) {
            // Guarantees the request settles even on OutOfMemoryError.
            return failure(FrameGrabCode.INTERNAL, "Unexpected native failure: $e")
        } finally {
            // Cleanup never replaces the original error, and never claims to have
            // rolled back an already committed destination.
            if (!committed) {
                runCatching { temporary.delete() }
            }
            runCatching { retriever?.release() }
        }
    }

    private fun openSource(
        retriever: MediaMetadataRetriever,
        source: FrameGrabSource,
    ) {
        try {
            when (source) {
                is FrameGrabSource.LocalFile -> retriever.setDataSource(source.file.absolutePath)
                is FrameGrabSource.Content -> {
                    val context =
                        NitroModules.applicationContext
                            ?: throw FrameGrabException(
                                FrameGrabCode.INTERNAL,
                                "No React application context available for `content://` access.",
                            )
                    retriever.setDataSource(context, source.uri)
                }
                is FrameGrabSource.Remote ->
                    retriever.setDataSource(source.uri, HashMap<String, String>())
            }
        } catch (e: FrameGrabException) {
            throw e
        } catch (e: Throwable) {
            throw FrameGrabException(
                FrameGrabCode.SOURCE_UNREADABLE,
                "Could not open source: ${e.javaClass.simpleName}",
            )
        }
    }

    private class Metadata(
        val orientedWidth: Int,
        val orientedHeight: Int,
        val rotationDegrees: Int,
    )

    private fun readMetadata(
        retriever: MediaMetadataRetriever,
        plan: Plan,
    ): Metadata {
        fun key(id: Int): String? =
            try {
                retriever.extractMetadata(id)
            } catch (e: Throwable) {
                throw FrameGrabException(
                    FrameGrabCode.SOURCE_UNREADABLE,
                    "Could not read source metadata: ${e.javaClass.simpleName}",
                )
            }

        if (key(MediaMetadataRetriever.METADATA_KEY_HAS_VIDEO) != "yes") {
            throw FrameGrabException(
                FrameGrabCode.SOURCE_UNREADABLE,
                "Source has no usable video track.",
            )
        }

        // These report the *encoded* dimensions; rotation is a separate key.
        val encodedWidth = key(MediaMetadataRetriever.METADATA_KEY_VIDEO_WIDTH)?.toIntOrNull()
        val encodedHeight = key(MediaMetadataRetriever.METADATA_KEY_VIDEO_HEIGHT)?.toIntOrNull()
        if (encodedWidth == null || encodedHeight == null ||
            encodedWidth <= 0 || encodedHeight <= 0
        ) {
            throw FrameGrabException(
                FrameGrabCode.SOURCE_UNREADABLE,
                "Source reports unusable video dimensions.",
            )
        }

        val rawRotation =
            key(MediaMetadataRetriever.METADATA_KEY_VIDEO_ROTATION)?.toIntOrNull() ?: 0
        val rotation = ((rawRotation % 360) + 360) % 360

        // A timestamp equal to the duration is out of range. An unknown duration
        // is not proof of anything: let extraction decide.
        val durationMs = key(MediaMetadataRetriever.METADATA_KEY_DURATION)?.toLongOrNull()
        if (durationMs != null && durationMs > 0 && durationMs < Long.MAX_VALUE / 1000) {
            if (plan.timeUs >= durationMs * 1000L) {
                throw FrameGrabException(
                    FrameGrabCode.TIMESTAMP_OUT_OF_RANGE,
                    "Requested ${plan.timeUs / 1000.0} ms but the source is $durationMs ms long.",
                )
            }
        }

        val swap = rotation == 90 || rotation == 270
        return Metadata(
            orientedWidth = if (swap) encodedHeight else encodedWidth,
            orientedHeight = if (swap) encodedWidth else encodedHeight,
            rotationDegrees = rotation,
        )
    }

    private fun extractFrame(
        retriever: MediaMetadataRetriever,
        timeUs: Long,
        option: Int,
        targetWidth: Int,
        targetHeight: Int,
    ): Bitmap {
        val bitmap =
            try {
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O_MR1) {
                    // Scaled extraction: the platform never hands us more pixels
                    // than we asked for. It is a bound, not a guarantee about how
                    // the decoder allocates internally.
                    retriever.getScaledFrameAtTime(timeUs, option, targetWidth, targetHeight)
                } else {
                    // minSdk is 23, so this branch is reachable. One bounded
                    // resize happens later, based on the bitmap we actually got.
                    retriever.getFrameAtTime(timeUs, option)
                }
            } catch (e: Throwable) {
                throw FrameGrabException(
                    FrameGrabCode.FRAME_EXTRACTION,
                    "Frame extraction failed: ${e.javaClass.simpleName}",
                )
            }

        // Never retry at full resolution: that amplifies memory pressure and
        // hides genuinely unsupported media.
        return bitmap
            ?: throw FrameGrabException(
                FrameGrabCode.FRAME_EXTRACTION,
                "No frame available at the requested timestamp.",
            )
    }

    // endregion

    // region Scaling / orientation

    /** Fits the oriented display size inside [maxWidth] without ever upscaling. */
    private fun fit(
        width: Int,
        height: Int,
        maxWidth: Int,
    ): Pair<Int, Int> {
        val scale = min(1.0, maxWidth.toDouble() / width.toDouble())
        val targetWidth = max(1, min((width * scale).roundToInt(), maxWidth))
        val targetHeight = max(1, (height * scale).roundToInt())
        if (targetWidth.toLong() * targetHeight > MAX_OUTPUT_AREA) {
            throw FrameGrabException(
                FrameGrabCode.INVALID_ARGUMENT,
                "Output would be ${targetWidth}x$targetHeight, over the $MAX_OUTPUT_AREA pixel limit.",
            )
        }
        return targetWidth to targetHeight
    }

    /**
     * `MediaMetadataRetriever` is expected to return display-oriented bitmaps.
     * We only correct the one case we can actually detect: a 90/270 source whose
     * returned bitmap still has the *encoded* aspect orientation.
     *
     * Dimensions cannot resolve 180-degree or square frames, so those are left to
     * the platform rather than guessed at. See README, "Android orientation".
     *
     * This rests on one AOSP assumption: `METADATA_KEY_VIDEO_WIDTH`/`_HEIGHT`
     * report *encoded* dimensions, with rotation carried separately in
     * `METADATA_KEY_VIDEO_ROTATION`. A device that pre-swapped those keys *and*
     * returned an oriented bitmap would be rotated twice here. That is what the
     * TOP/LEFT fixture set in `example/` exists to catch; run it on unfamiliar
     * hardware before trusting rotated output.
     */
    private fun rotateIfPlatformDidNot(
        bitmap: Bitmap,
        rotationDegrees: Int,
        orientedWidth: Int,
        orientedHeight: Int,
    ): Bitmap {
        if (rotationDegrees != 90 && rotationDegrees != 270) return bitmap
        if (bitmap.width == bitmap.height || orientedWidth == orientedHeight) return bitmap
        if ((bitmap.width > bitmap.height) == (orientedWidth > orientedHeight)) return bitmap

        val matrix = Matrix().apply { postRotate(rotationDegrees.toFloat()) }
        return Bitmap.createBitmap(bitmap, 0, 0, bitmap.width, bitmap.height, matrix, true)
    }

    /** Single bounded, aspect-preserving resize based on the bitmap we actually got. */
    private fun boundWidth(
        bitmap: Bitmap,
        maxWidth: Int,
    ): Bitmap {
        if (bitmap.width <= maxWidth) return bitmap
        val scale = maxWidth.toDouble() / bitmap.width.toDouble()
        val height = max(1, (bitmap.height * scale).roundToInt())
        return Bitmap.createScaledBitmap(bitmap, maxWidth, height, true)
    }

    // endregion

    private fun elapsedMs(startNanos: Long): Double = (System.nanoTime() - startNanos) / 1_000_000.0
}

private fun failure(
    code: String,
    message: String,
): NativeFrameGrabOutcome =
    NativeFrameGrabOutcome(
        result = null,
        error = NativeFrameGrabFailure(code = code, message = message),
    )
